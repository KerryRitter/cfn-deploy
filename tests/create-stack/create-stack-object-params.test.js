const AWS = require('aws-sdk-mock');
const lib = require('../..');


describe('cfn-deploy', () => {
  beforeAll(() => {
    let isChangeSetCreated = false;

    AWS.mock('CloudFormation', 'validateTemplate', (params, callback) => {
      callback(null, {});
    });
    AWS.mock('CloudFormation', 'describeStacks', (params, callback) => {
      if (!isChangeSetCreated) {
        callback();
      } else {
        // Stack exists once changeset has been created
        callback(null, {
          Stacks: [
            {
              StackStatus: 'REVIEW_IN_PROGRESS',
            },
          ],
        });
      }
    });
    AWS.mock('CloudFormation', 'createChangeSet', (params, callback) => {
      expect(params).toMatchSnapshot();
      isChangeSetCreated = true;
      callback(null, {
        Id: 'fake-changeset-id',
      });
    });
    AWS.mock('CloudFormation', 'describeChangeSet', (params, callback) => {
      callback(null, {
        ChangeSetId: 'fake-changeset-id',
      });
    });
    AWS.mock('CloudFormation', 'executeChangeSet', (params, callback) => {
      callback();
    });
    AWS.mock('CloudFormation', 'deleteChangeSet', (params, callback) => {
      callback(new Error('SHOULD_NOT_BE_CALLED'));
    });
    AWS.mock('CloudFormation', 'waitFor', (event, params, callback) => {
      if (event === 'changeSetCreateComplete') {
        callback(null, {
          ChangeSetId: 'fake-changeset-id',
        });
      } else if (event === 'stackCreateComplete') {
        callback(null, {
          Stacks: [
            {
              StackStatus: 'CREATE_COMPLETE',
            },
          ],
        });
      }
    });
  });

  it('should successfully create new stack', (done) => {
    const events = lib({
      region: 'us-east-1',
      stackName: 'new-stack',
      template: './tests/templates/params-template.yaml',
      parameters: {
        S3BucketName: 'from-object-params',
        DummyParam: 'dummy-param',
        OverwritableParam: 'from-object-params',
        ParamSource: 'plain-object',
      },
    });
    events.on('ERROR', err => done(err));
    events.on('COMPLETE', (data) => {
      expect(data).toMatchSnapshot();
    });
    events.on('FINALLY', done);
  });

  afterAll(() => {
    AWS.restore('CloudFormation');
  });
});
