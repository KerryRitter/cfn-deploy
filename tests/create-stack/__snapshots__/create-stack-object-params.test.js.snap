// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cfn-deploy should successfully create new stack 1`] = `
Object {
  "Capabilities": undefined,
  "ChangeSetName": "cfn-deploy",
  "ChangeSetType": "CREATE",
  "Description": "Created with cfn-deploy",
  "Parameters": Array [
    Object {
      "ParameterKey": "S3BucketName",
      "ParameterValue": "from-object-params",
    },
    Object {
      "ParameterKey": "DummyParam",
      "ParameterValue": "dummy-param",
    },
    Object {
      "ParameterKey": "OverwritableParam",
      "ParameterValue": "from-object-params",
    },
    Object {
      "ParameterKey": "ParamSource",
      "ParameterValue": "plain-object",
    },
  ],
  "StackName": "new-stack",
  "TemplateBody": "Parameters:
  S3BucketName:
    Description: \\"S3 Bucket name\\"
    Type: String
  DummyParam:
    Description: \\"Dummy parameter\\"
    Type: String

Resources:
  S3Bucket:
    Type: AWS::S3::Bucket
    Properties:
      BucketName: !Sub \\"\${AWS::StackName}-\${S3BucketName}\\"
",
}
`;

exports[`cfn-deploy should successfully create new stack 2`] = `
Object {
  "type": "UPDATE",
}
`;
